File: Java/Fundamentals/Platform/Concurrency/ThreadCoordTest.java
class ThreadCoordTest{

	private static volatile int value = 0;
	private static Object coordinator = new Object();

	private static void consume() throws InterruptedException{
		System.out.printf("Consumer<%d> ready...%n", Thread.currentThread().getId());
		//while(value == 0) Thread.yield();
		synchronized(coordinator){
			coordinator.wait();
			System.out.printf("Processed value = %d%n", value * value);
		}
	}

	private static void produce(){
		System.out.printf("Producer<%d> ready...%n", Thread.currentThread().getId());
		int result = Worker.doWork();
		synchronized(coordinator){
			value = result;
			coordinator.notify();
		}
	}

	public static void main(String[] args) throws Exception{
		Thread child = new Thread(ThreadCoordTest::produce);
		child.start();
		consume();
	}
}

File: C/UNIX/Concurrency/threadcoordtest.c
#include "worker.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>

volatile long value = 0;

void Produce(void)
{
	printf("Producer <%x> ready...\n", pthread_self());
	value = DoWork(0);
}

void* Consume(void* arg)
{
	printf("Consumer <%x> ready...\n", pthread_self());
	while(value == 0)
		pthread_yield();
	printf("Processed value = %ld\n", value * value);
}

int main(void)
{
	pthread_t child;

	pthread_create(&child, NULL, Consume, NULL);
	Produce();

	pthread_join(child, NULL);
}













